File: AI_LAB1/EX1/Nhom02_29_B2113306_KimKhanhDang_BaiTap2_DFS_DongNuoc.py
# tankcapacityX = 9 
# tankcapacityY = 4 
# empty = 0
# goal = 6 
# Maxlength = 100  

# action = [
#     "First State",
#     "pour Water Full X",
#     "pour Water Full Y",
#     "pour Water Empty X",
#     "pour Water Empty Y",
#     "pour Water X to Y",
#     "pour Water Y to X"
# ]

# class State:
#     def __init__(self, x, y):
#         self.x = x
#         self.y = y

# def makeNullState():
#     return State(0, 0)

# def printState(state):
#     print("\n X:", state.x, "--- Y:", state.y)

# def goalCheck(state):
#     return state.x == goal or state.y == goal

# def pourWaterFullX(curState, result):
#     if curState.x < tankcapacityX:
#         result.x = tankcapacityX
#         result.y = curState.y
#         return 1
#     return 0

# def pourWaterFullY(curState, result):
#     if curState.y < tankcapacityY:
#         result.y = tankcapacityY
#         result.x = curState.x
#         return 1
#     return 0


# def pourWaterEmptyX(curState, result):
#     if curState.x > 0:
#         result.x = 0
#         result.y = curState.y
#         return 1
#     return 0


# def pourWaterEmptyY(curState, result):
#     if curState.y > 0:
#         result.y = 0
#         result.x = curState.x
#         return 1
#     return 0


# def pourWaterXY(curState, result):
#     if curState.x > 0 and curState.y < tankcapacityY:
#         result.x = max(curState.x - (tankcapacityY - curState.y), empty)
#         result.y = min(curState.y + curState.x, tankcapacityY)
#         return 1
#     return 0


# def pourWaterYX(curState, result):
#     if curState.y > 0 and curState.x < tankcapacityX:
#         result.x = min(curState.x + curState.y, tankcapacityX)
#         result.y = max(curState.y - (tankcapacityX - curState.x), empty)
#         return 1
#     return 0


# def callOperator(curState, result, option):
#     if option == 1:
#         return pourWaterFullX(curState, result)
#     elif option == 2:
#         return pourWaterFullY(curState, result)
#     elif option == 3:
#         return pourWaterEmptyX(curState, result)
#     elif option == 4:
#         return pourWaterEmptyY(curState, result)
#     elif option == 5:
#         return pourWaterXY(curState, result)
#     elif option == 6:
#         return pourWaterYX(curState, result)
#     else:
#         print("\n--- Error calls operators ---")
#         return 0


# class Node:
#     def __init__(self, state, parent, noFunction):
#         self.state = state
#         self.parent = parent
#         self.noFunction = noFunction


# def compareStates(a, b):
#     return a.x == b.x and a.y == b.y


# def findState(state, stack):
#     for node in stack:
#         if compareStates(node.state, state):
#             return 1
#     return 0



# def DFS_Algorithm(state):
#     openStack, closeStack = [], []
#     root = Node(state, None, 0)
#     openStack.append(root)

#     while len(openStack):
#         node = openStack[-1]
#         openStack.pop() #openQueue.pop(0)-- BFS
#         closeStack.append(node)
#         if goalCheck(node.state):
#             return node
#         for opt in range(1, 7):
#             newState = makeNullState()
#             if callOperator(node.state, newState, opt):
#                 if findState(newState, closeStack) or findState(newState, openStack):
#                     continue
#                 newNode = Node(newState, node, opt)
#                 openStack.append(newNode)
#     return None

# def printWaysToGetGoal(node):
#     stackPrint = []
#     while node is not None:
#         stackPrint.append(node)
#         node = node.parent

#     noAction = 0
#     while len(stackPrint):
#         print("\nAction", noAction, ":", action[stackPrint[-1].noFunction])
#         printState(stackPrint[-1].state)
#         stackPrint.pop()
#         noAction += 1


# def main():
#     curState = State(0, 0)
#     p = DFS_Algorithm(curState)
#     printWaysToGetGoal(p)


# if __name__ == "__main__":
#     main()

TANK_CAPACITY_X = 9
TANK_CAPACITY_Y = 4
GOAL = 6

ACTIONS = [
    "Initial State",
    "Fill X",
    "Fill Y",
    "Empty X",
    "Empty Y",
    "Pour X to Y",
    "Pour Y to X"
]

class State:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def pour_water(source, target, cur_state):
    max_pour = min(cur_state[source], TANK_CAPACITY_Y - cur_state[target])
    new_state = cur_state[:]
    new_state[source] -= max_pour
    new_state[target] += max_pour
    return State(*new_state)

POUR_OPERATORS = [
    lambda state: State(TANK_CAPACITY_X, state.y),                  # Fill X
    lambda state: State(state.x, TANK_CAPACITY_Y),                  # Fill Y
    lambda state: State(0, state.y),                                 # Empty X
    lambda state: State(state.x, 0),                                 # Empty Y
    lambda state: pour_water(0, 1, state),                          # Pour X to Y
    lambda state: pour_water(1, 0, state)                           # Pour Y to X
]

class Node:
    def __init__(self, state, parent, action_index):
        self.state = state
        self.parent = parent
        self.action_index = action_index

def dfs_algorithm(initial_state):
    open_stack, close_stack = [], []
    root = Node(initial_state, None, 0)
    open_stack.append(root)

    while open_stack:
        node = open_stack.pop()
        close_stack.append(node)
        if node.state.x == GOAL or node.state.y == GOAL:
            return node
        for action_index, operator in enumerate(POUR_OPERATORS, start=1):
            new_state = operator(node.state)
            if new_state in [n.state for n in close_stack]:
                continue
            new_node = Node(new_state, node, action_index)
            open_stack.append(new_node)
    return None

def print_ways_to_get_goal(node):
    stack_print = []
    while node is not None:
        stack_print.append(node)
        node = node.parent

    action_count = 0
    while stack_print:
        print(f"Action {action_count}: {ACTIONS[stack_print[-1].action_index]}")
        print(f"X: {stack_print[-1].state.x} --- Y: {stack_print[-1].state.y}")
        stack_print.pop()
        action_count += 1

def main():
    initial_state = State(0, 0)
    goal_node = dfs_algorithm(initial_state)
    if goal_node:
        print_ways_to_get_goal(goal_node)
    else:
        print("No solution found.")

if __name__ == "__main__":
    main()

File: AI_LAB1/EX1/Nhom02_29_B2113306_KimKhanhDang_BaiTap2_DFS_DongNuoc.c
#include <stdio.h>
#include <stdlib.h>
#define tankcapacity_X 9
#define tankcapacity_Y 4
#define empty 0
#define goal 6
#define Maxlength 100

typedef struct{
	int x;
	int y;
}State;

void makeNullState(State *state){
	state->x = 0;
	state->y = 0;
}

void print_State(State state){
	printf("\n   X:%d --- Y:%d", state.x, state.y);
}

int goalcheck(State state){
	return (state.x == goal || state.y == goal);
}

int max(int a, int b){
	return a > b ? a:b;
}

int min(int a, int b){
	return a < b ? a:b;
}

int pourWaterFullX(State cur_state, State *result){
	if(cur_state.x < tankcapacity_X){
		result->x = tankcapacity_X;
		result->y = cur_state.y;
		return 1;
	}
	return 0;
}

int pourWaterFullY(State cur_state, State *result){
	if(cur_state.y < tankcapacity_Y){
		result->y = tankcapacity_Y;
		result->x = cur_state.x;
		return 1;
	}
	return 0;
}

int pourWaterEmptyX(State cur_state, State *result){
	if(cur_state.x > 0){
		result->x = empty;
		result->y = cur_state.y;
		return 1;
	}
	return 0;
}

int pourWaterEmptyY(State cur_state, State *result){
	if(cur_state.y > 0){
		result->y = empty;
		result->x = cur_state.x;
		return 1;
	}
	return 0;
}

int pourWaterXY(State cur_state, State *result){
	if(cur_state.x > 0 && cur_state.y < tankcapacity_Y){
		result->x = max(cur_state.x - (tankcapacity_Y - cur_state.y), empty);
		result->y = min(cur_state.x + cur_state.y, tankcapacity_Y);
		return 1;
	}
	return 0;
}

int pourWaterYX(State cur_state, State *result){
	if(cur_state.y > 0 && cur_state.x < tankcapacity_X){
		result->y = max(cur_state.y - (tankcapacity_X - cur_state.x), empty);
		result->x = min(cur_state.y + cur_state.x, tankcapacity_X);
		return 1;
	}
	return 0;
}

int call_operator(State cur_state, State *result, int option){
	switch(option){
		case 1: return pourWaterFullX(cur_state, result);
		case 2: return pourWaterFullY(cur_state, result);
		case 3: return pourWaterEmptyX(cur_state, result);
		case 4: return pourWaterEmptyY(cur_state, result);
		case 5: return pourWaterXY(cur_state, result);
		case 6: return pourWaterYX(cur_state, result);
		default: printf("Error calls operators");
			return 0;
	}
}

const char* action[] = {"First State", "pour water full X", "pour water full Y", 
						"pour water empty X", "pour water empty Y", "pour water X to Y", "pour water Y to X"};


typedef struct Node{
	State state;
	struct Node* Parent;
	int no_func;
}Node;

typedef struct{
	Node* Element[Maxlength];
	int Top_idx;
}Stack;

void makeNull_Stack(Stack *stack){
	stack->Top_idx = Maxlength;
}

int empty_Stack(Stack stack){
	return stack.Top_idx == Maxlength;
}

int full_Stack(Stack stack){
	return stack.Top_idx == 0;
}

int compareStates(State openStack, State state){
	if(openStack.x == state.x && openStack.y == state.y)
		return 1;
	return 0;
}

void push(Node* x, Stack *stack){
	if(full_Stack(*stack))
		printf("Error!Stack is full");
	else{
		stack->Top_idx -=1;
		stack->Element[stack->Top_idx-1] = x;
	}
}

Node* top(Stack stack){
	if(!empty_Stack(stack)){
		return stack.Element[stack.Top_idx-1];
	}
	return NULL;
}

void pop(Stack *stack){
	if(!empty_Stack(*stack))
		stack->Top_idx += 1;
	else
		printf("Error!Stack is empty");
}

int find_State(State state, Stack openStack){
	while(!empty_Stack(openStack)){
		if(compareStates(top(openStack)->state, state))
			return 1;
		pop(&openStack);
	}
	return 0;
}

Node* DFS_Algorithm(State state){
	Stack Open_DFS;
	Stack Close_DFS;
	makeNull_Stack(&Open_DFS);
	makeNull_Stack(&Close_DFS);
	Node* root = (Node*)malloc(sizeof(Node));
	root->state = state;
	root->Parent = NULL;
	root->no_func = 0;
	push(root, &Open_DFS);
	while(!empty_Stack(Open_DFS)){
		Node* node = top(Open_DFS);
		pop(&Open_DFS);
		push(node, &Close_DFS);
		if(goalcheck(node->state))
			return node;
		int opt;
		for(opt=1; opt<=6; opt++){
			State newstate;
			makeNullState(&newstate);
			if(call_operator(node->state, &newstate, opt)){
				if(find_State(newstate, Close_DFS) || find_State(newstate, Open_DFS))
					continue;
				Node* newNode = (Node*)malloc(sizeof(Node));
				newNode->state = newstate;
				newNode->Parent = node;
				newNode->no_func = opt;
				push(newNode, &Open_DFS);
			}
		}
	}
	free(root);
	return NULL;
}

void print_WayToGetGoal(Node* node){
	Stack stackPrint;
	makeNull_Stack(&stackPrint);
	while(node->Parent != NULL){
		push(node, &stackPrint);
		node = node->Parent;
	}
	push(node, &stackPrint);
	int no_act = 0;
	while(!empty_Stack(stackPrint)){
		printf("\nAction: %d: %s", no_act, action[top(stackPrint)->no_func]);
		print_State(top(stackPrint)->state);
		pop(&stackPrint);
		no_act++;
	}
}


int main(){
	State cur_state = {0, 0};
	Node* p = DFS_Algorithm(cur_state);
	print_WayToGetGoal(p);
	return 0;
}


